### PR TITLE
ExternalXMLLoaders reload with server configuration support

### LINKED ISSUE #28142: Add support for relative paths to load external machine learning models.
> You have to provide the following information whenever possible.

Relative paths for ml models works wrong

> A clear and concise description of what works not as it is supposed to.

It's possibly broken here https://github.com/ClickHouse/ClickHouse/commit/cc65a1db972973470c282e0795818af9e0a8c0a9
In [this](https://github.com/ClickHouse/ClickHouse/blob/master/src/Interpreters/ExternalLoaderXMLConfigRepository.cpp#L39) string there is always returned default value, so if the config path is overridden we got the wrong path for the models.

**Does it reproduce on recent release?**

Yes

**How to reproduce**

Use clickhouse version 21.5>
Set a relative path to ml models in config.xml
User --config arg on server startup
Place ml models to the target directory
`SELECT COUNT(*) FROM system.models` remains zero

**Expected behavior**
As in exteranal dictionary loader

### PR DESCRIPTION
**Pull request by kitaisreal**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Reload dictionaries, models, user defined executable functions if servers config `dictionaries_config`, `models_config`, `user_defined_executable_functions_config` changes. Closes #28142.

### CODE REVIEW

File: programs/server/Server.cpp [hunk 1]
```diff
@@ -1469,7 +1472,29 @@ if (ThreadFuzzer::instance().isEffective())
         /// try to load dictionaries immediately, throw on error and die
         try
         {
-            global_context->loadDictionaries(config());
+            global_context->loadOrReloadDictionaries(config());
+        }
+        catch (...)
+        {
+            LOG_ERROR(log, "Caught exception while loading dictionaries.");
```
Comment (nikitamikhaylov) [inline programs/server/Server.cpp:1479]: tryLogCurrentException? Because LOG_ERROR may throw an exception

File: src/Interpreters/Context.cpp [hunk 2]
```diff
@@ -1361,20 +1376,29 @@ ExternalModelsLoader & Context::getExternalModelsLoaderUnlocked()
...
 
-void Context::setExternalModelsConfig(const ConfigurationPtr & config, const std::string & config_name)
+void Context::loadOrReloadModels(const Poco::Util::AbstractConfiguration & config)
 {
+    auto patterns_values = getMultipleValuesFromConfig(config, "", "models_config");
+    std::unordered_set<std::string> patterns(patterns_values.begin(), patterns_values.end());
+
     std::lock_guard lock(shared->external_models_mutex);
 
-    if (shared->external_models_config && isSameConfigurationWithMultipleKeys(*config, *shared->external_models_config, "", config_name))
```
Comment (Avogar) [inline src/Interpreters/Context.cpp:1368]: I see you removed logic when we don't reload models config if it didn't change, why?

**Comment by Avogar** [comment]
Will you add an integration test on it? Like existed test `test_catboost_model_config_reload`

File: src/Interpreters/Context.cpp (hunk 2 above)
Comment (kitaisreal) [reply src/Interpreters/Context.cpp:1368]: We update patterns in repository, then ExternalLoader will check and remove old models, load new models if necessary.

File: programs/server/Server.cpp (hunk 1 above)
Comment (kitaisreal) [reply programs/server/Server.cpp:1479]: Fixed.

File: src/Interpreters/ExternalLoaderXMLConfigRepository.cpp [hunk 3]
```diff
@@ -1,20 +1,27 @@
 #include <Interpreters/ExternalLoaderXMLConfigRepository.h>
 
+#include <filesystem>
+#include <iostream>
```
Comment (Avogar) [inline src/Interpreters/ExternalLoaderXMLConfigRepository.cpp:4]: Unused header?

Comment (kitaisreal) [reply src/Interpreters/ExternalLoaderXMLConfigRepository.cpp]: Fixed.

**Pull request merged**